### PR TITLE
Jamie/4366 disappearing pagination

### PR DESCRIPTION
**Pull request by SEAjamieD**
### :nut_and_bolt: Description: What code changed, and why?
A customer reported that their pagination was disappearing/closing their sidebar on client runs > run history.  This turned out to be 2 different bugs.

1.  When there was only 1 selectable page left to view, the pagination component would disappear.  For example, if a user has 11 pages of content, on a default pagination component we show 10 selectable pages at a time.  When they get to page 11 - there is only 1 page available, so the page hides.  In this case, it would also cause the sidebar to close.

2.  The selectable pages were being calculated off the amount of selectable content, in this particular case - when the content is filtered, the page picker didn't know how to figure out what the current page was, so it was calculating against pages that didn't exist, which would lead the user to be able to select pages that had no content, which would then cause the pagination component to disappear - and in this case, the sidebar to close.

### :chains: Related Resources
Fixes: https://github.com/chef/automate/issues/4366

### :+1: Definition of Done
Pagination component works as expected.  
1. Shows when there is more than one page of total selectable content.
2. Resets when filters are added.
3. Shows the current page after reset has been done.

### :athletic_shoe: How to Build and Test the Change
cd Automate
`build components/automate-ui-devproxy && start_all_services`

cd Automate/components/automate-ui
`make serve`

in Hab Studio, load up a node with lots of client run history
`for i in {1..105}; do chef_load_nodes 1; done`

navigate to https://a2-dev.test/infrastructure/client-runs
if a fresh install, click on the only node available
if not a fresh install, likely search for the node named `chef-load-1`

on the single node page, click on the run history button in the top right corner.
navigate through the pagination on the bottom of the run history slideout.
use the filters on the top to change the pagination.

### :white_check_mark: Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] Tests added/updated?
- [ ] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable
Client Run history pagination

### CODE REVIEW

File: components/automate-ui/src/app/components/page-picker/page-picker.component.html [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<ng-container *ngIf="selectablePages.length > 1">
+<ng-container *ngIf="selectablePages.length > 1 || last > 1">
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/components/page-picker/page-picker.component.html:1]: It's possible that sometimes we have a scenario with only 1 selectable page, however we're looking at say...page 11.  This makes sure that if we have only one selectable page, but more than 1 page in total, we still see the pagination component.

@msorens - thinking now that I probably need to pipe this, huh?

File: components/automate-ui/src/app/page-components/run-history/run-history.component.scss [hunk 2]
```diff
@@ -180,6 +180,5 @@ app-date-selector {
 }
 
 .history-pagination {
-  margin-bottom: 1em;
-  justify-content: center;
+  padding: 1em;
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/page-components/run-history/run-history.component.scss:183]: Small styling update - gives the pagination component a little breathing room.

File: components/automate-ui/src/app/components/page-picker/page-picker.component.ts [hunk 3]
```diff
@@ -54,6 +54,7 @@ export class PagePickerComponent implements OnChanges {
 
   ngOnChanges() {
     this.last = Math.ceil(this.total / this.perPage) || 1;
+    this.page = this.page > this.last ? this.last : this.page;
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/components/page-picker/page-picker.component.ts:57]: This was the main culprit, the current page data wasn't being updated, which was throwing off the math below creating selectable pages that didn't actually exist.  This makes sure that if we're dropping to a fewer amount of pages, the current page is adjusted to reflect that change.

File: components/automate-ui/src/app/components/page-picker/page-picker.component.html (hunk 1 above)
Comment (msorens) [reply components/automate-ui/src/app/components/page-picker/page-picker.component.html:1]: No, using properties in expressions local to the template like this is fine; it is only when you introduce a function call that a pipe would improve performance.

File: components/automate-ui/src/app/components/page-picker/page-picker.component.ts (hunk 3 above)
Comment (msorens) [reply components/automate-ui/src/app/components/page-picker/page-picker.component.ts:57]: Good, but can simplify a bit:
```suggestion
    this.page = Math.min(this.page, this.last);
```

**Review by msorens — COMMENTED** [review]
I am not quite following your "case 1". You indicate there is a default of 10 pages but I am only seeing 3 at a time (and your video also shows only 3). And I could not get the picker to show only page 11 by itself in order to test the fix for case 1.

**Comment by SEAjamieD** [comment]
> I am not quite following your "case 1". You indicate there is a default of 10 pages but I am only seeing 3 at a time (and your video also shows only 3). And I could not get the picker to show only page 11 by itself in order to test the fix for case 1.

@msorens - I was speaking in general about the page-picker component.  By default it sets the amount of selectable pages as 10, however in this specific case, we have it set to max out at 3 pages.  You can see the issue if you remove the `maxSelectablePages` property in the `run-history.html` page.  

after looking at your comment above, I updated the property to be named `maxSelectablePages` so that it was more easily understandable.
bug

**Comment by msorens** [comment]
Thanks for the additional details, @SEAjamieD. Looks one a minor issue: when I revert to the default of 10 on your branch I am seeing this--note that the page 11 icon is disabled and it should not be:
image

**Comment by SEAjamieD** [comment]
> Thanks for the additional details, @SEAjamieD. Looks one a minor issue: when I revert to the default of 10 on your branch I am seeing this--note that the page 11 icon is disabled and it should not be:
> image

@msorens - Really fantastic catch there - thank you!

**Review by Shadae — APPROVED** [review]
:+1:

File: components/automate-ui/src/app/components/page-picker/page-picker.component.html [hunk 4]
```diff
@@ -17,7 +17,7 @@
       class="page-picker-item"
       [attr.primary]="item === page"
       [attr.secondary]="item !== page"
-      [attr.disabled]="selectablePages.length <= 1"
+      [attr.disabled]="selectablePages.length <= 1 && last <= 1"
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/components/page-picker/page-picker.component.html:20]: Similar to the addition above, this makes sure that the page selector button is _not_ disabled when we hit a scenario where there is only one selectable page showing, but may still have multiple pages.

Comment (Shadae) [reply components/automate-ui/src/app/components/page-picker/page-picker.component.html:20]: Nice!

File: components/automate-ui/src/app/components/page-picker/page-picker.component.ts [hunk 5]
```diff
@@ -54,24 +54,17 @@ export class PagePickerComponent implements OnChanges {
...
-
-    for (let pageIndex = selectedIndex; pageIndex < this.maxPageItems; ++pageIndex) {
-      const next = this.page + (pageIndex - selectedIndex);
-      if (next <= this.last) {
-        pages[pageIndex] = next;
-      }
+    for (let pageIndex = 0, pg = this.page - selectedIndex;
+      pageIndex < this.maxSelectablePages && pg <= this.last;
+      ++pageIndex, pg = this.page + pageIndex - selectedIndex) {
+      pages[pageIndex] = pg;
```
Comment (msorens) [inline components/automate-ui/src/app/components/page-picker/page-picker.component.ts:67]: Trying to understand all the nuances and edges of the original code was frazzling my brain; just had to tidy it up here (and add some test cases proving it is equivalent!).

**Review by msorens — APPROVED** [review]
Working as advertised!
While I was making sure I understood the code I took the liberty of tidying up some nearby calculations; hope you do not mind.

File: components/automate-ui/src/app/components/page-picker/page-picker.component.ts (hunk 5 above)
Comment (SEAjamieD) [reply components/automate-ui/src/app/components/page-picker/page-picker.component.ts:67]: amazing cleanup, thanks! 🎉   And tests always apreciated!

**Pull request merged**